log4j-api/src/main/java/org/apache/logging/log4j/status/StatusLogger.java

```diff
@@ -16,311 +16,567 @@
+                // There is nothing we can do but simply dumping the failure.
+                error.printStackTrace(System.err);
+            }
+            return properties;
+        }
+    }
+
+    private static volatile StatusLogger INSTANCE = new StatusLogger();
+
+    private final Config config;
+
+    private final StatusConsoleListener defaultListener;
+
+    private final Collection<StatusListener> listeners;
+
+    private final transient ReadWriteLock listenerLock = new ReentrantReadWriteLock();
 
-    private final Queue<StatusData> messages = new BoundedQueue<>(MAX_ENTRIES);
+    private final transient Lock listenerReadLock = listenerLock.readLock();
 
-    @SuppressWarnings("NonSerializableFieldInSerializableClass")
-    // ReentrantLock is Serializable
-    private final Lock msgLock = new ReentrantLock();
+    private final transient Lock listenerWriteLock = listenerLock.writeLock();
 
-    private int listenersLevel;
+    private final Queue<StatusData> buffer = new ConcurrentLinkedQueue<>();
 
     /**
-     * Constructs the singleton instance for the STATUS_LOGGER constant.
-     * <p>
-     * This is now the logger level is set:
-     * </p>
-     * <ol>
-     * <li>If the property {@value Constants#LOG4J2_DEBUG} is {@code "true"}, then use {@link Level#TRACE}, otherwise,</li>
-     * <li>Use {@link Level#ERROR}</li>
-     * </ol>
-     * <p>
-     * This is now the listener level is set:
-     * </p>
-     * <ol>
-     * <li>If the property {@value #DEFAULT_STATUS_LISTENER_LEVEL} is set, then use <em>it</em>, otherwise,</li>
-     * <li>Use {@link Level#WARN}</li>
-     * </ol>
-     * <p>
-     * See:
-     * <ol>
-     * <li>LOG4J2-1813 Provide shorter and more intuitive way to switch on Log4j internal debug logging. If system property
-     * "log4j2.debug" is defined, print all status logging.</li>
-     * <li>LOG4J2-3340 StatusLogger's log Level cannot be changed as advertised.</li>
-     * </ol>
-     * </p>
+     * Constructs the default instance.
+     * <b>Users should not create new instances, but use {@link #getLogger()} instead!</b>
      *
-     * @param name The logger name.
-     * @param messageFactory The message factory.
+     * @since 2.23.0
      */
-    private StatusLogger(
-            final String name, final MessageFactory messageFactory, final SimpleLoggerFactory loggerFactory) {
-        super(name, messageFactory);
-        final Level loggerLevel = DEBUG_ENABLED ? Level.TRACE : Level.ERROR;
-        this.logger = loggerFactory.createSimpleLogger("StatusLogger", loggerLevel, messageFactory, System.err);
-        this.listenersLevel = Level.toLevel(DEFAULT_STATUS_LEVEL, Level.WARN).intLevel();
+    public StatusLogger() {
+        this(
+                StatusLogger.class.getSimpleName(),
+                ParameterizedNoReferenceMessageFactory.INSTANCE,
+                Config.getInstance(),
+                new StatusConsoleListener(Config.getInstance().defaultListenerLevel));
```

Comment:
> Users should not create new instances, but use {@link #getLogger()} instead!

I would remove this constructor instead and inline it in the only caller.

In our code we will rather use the other constructor.

Reply:
No-arg ctor allows restore once the state is corrupted. Consider the following:
```
StatusLogger.setLogger(mockStatusLogger);
// Okay, I am done with my tests.
// How can I restore the `StatusLogger`?
StatusLogger.setLogger(new StatusLogger());
```
One can argue that `StatusLogger.getLogger()` should be safely stored to a reference before being replaced.

Anyway... This was sort of my reasoning. I am not strongly opinionated. @ppkarwasz, what do you prefer?

Reply:
In my experience every `public` method **will** be used by users. So my strategy is to expose only those methods that are necessary or that we are willing to support.

Mocking the status logger is a very nice use case, but it will probably used in a single test class, because it is not parallelisable. All the other test classes should register a separate listener (like `JUnitListStatusListener`) or even better: inject a `Logger/StatusLogger` into the components they are testing.

Reply:
Removed the `StatusLogger()` ctor in f1e3e2f3238fbdd196b6bafe22451e10353690b2.